{graph, model, examples}: implement token tailoring functionality with strategies and counters

Rememorio opened the pull request:
- Expanded comments in the MiddleOutStrategy implementation to explain the positional bias in large context LLMs and the rationale for removing mid-sequence items first.
- Included references to relevant literature on attention bias and its implications for token tailoring strategies.
- Emphasized the heuristic nature of the strategy and suggested pairing with accurate token accounting methods for better results.

Rememorio committed:
Implement token tailoring functionality with strategies and counters

- Added `TokenCounter` and `TailoringStrategy` interfaces to the model package for managing token counts and tailoring messages.
- Introduced `SimpleTokenCounter`, `MiddleOutStrategy`, `HeadOutStrategy`, and `TailOutStrategy` implementations for token counting and message tailoring.
- Enhanced `LLMNodeFunc` and `Model` to support token tailoring, allowing messages to be trimmed based on token limits before processing.
- Added unit tests for token counting and tailoring strategies to ensure correct functionality.
- Created new test files for `token_tailor` and `tiktoken` to validate the new features and ensure compatibility with existing models.

Rememorio committed:
Refactor token tailoring implementation in LLM and OpenAI models

- Moved token tailoring logic into a dedicated method `applyTokenTailoring` in both `llmRunner` and `Model` to improve code organization and readability.
- Simplified the execution flow by removing redundant checks and consolidating token tailoring functionality.
- Enhanced comments for clarity on the purpose and behavior of the token tailoring process.

Rememorio committed:
Add interactive token tailoring example and documentation

- Introduced a new example in `examples/tailor/main.go` demonstrating interactive token tailoring using command line interface.
- Implemented features for message handling, token counting, and various tailoring strategies.
- Added a README in `examples/tailor/README.md` detailing usage, prerequisites, and commands for the interactive demo.
- Enhanced user experience by allowing bulk message addition and displaying message count during interaction.

Rememorio committed:
Add MiddleOutStrategy documentation to clarify token removal rationale

- Expanded comments in the MiddleOutStrategy implementation to explain the positional bias in large context LLMs and the rationale for removing mid-sequence items first.
- Included references to relevant literature on attention bias and its implications for token tailoring strategies.
- Emphasized the heuristic nature of the strategy and suggested pairing with accurate token accounting methods for better results.

Rememorio committed:
Add token counter and streaming support in interactive demo

- Introduced a new `tiktoken` counter option for improved token management in `examples/tailor/main.go`.
- Enhanced command-line flags to include a streaming option for real-time response handling.
- Updated README to reflect changes in prerequisites, usage instructions, and command outputs for better clarity.
- Added a summary feature to display tailored messages after processing, improving user experience during interaction.

Rememorio commented on model/token_tailor.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+// TailorMessages implements middle-out trimming.
+func (s *MiddleOutStrategy) TailorMessages(ctx context.Context, messages []Message, maxTokens int) ([]Message, error) {
```
这里的时间复杂度有点问题，我再想想

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
Refactor token counting and tailoring strategies for improved efficiency

- Updated `TokenCounter` interface to support counting tokens for individual messages and ranges, enhancing performance.
- Refactored `MiddleOutStrategy`, `HeadOutStrategy`, and `TailOutStrategy` to utilize prefix sums for efficient token management.
- Simplified message preservation logic to ensure system message and last turn are retained during tailoring.
- Enhanced unit tests to cover new functionality and validate performance improvements.
- Updated example and documentation to reflect changes in command-line flags and usage instructions.

Rememorio commented on examples/tailor/go.mod:
```diff
@@ -0,0 +1,26 @@
+module trpc.group/trpc-go/trpc-agent-go/examples/tailor
+
+go 1.23
```
这里做 `go.mod` 是因为 [tiktoken](https://github.com/tiktoken-go/tokenizer) 这个库最新版支持了 GPT-5，go 版本为 1.23，所以默认提供一个 `SimpleTokenCounter` 策略，按 4 字符 ≈ 1 token 粗略估计。需要精确估计的话可以使用 tiktoken 的 tokenizer。

Rememorio committed:
Update dependencies in go.mod and go.sum

- Updated `github.com/stretchr/testify` from v1.10.0 to v1.11.1 for improved testing capabilities.
- Added a replace directive for local development of `trpc.group/trpc-go/trpc-agent-go`.

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
refactor for defaults

Refactor token tailoring implementation to use separate options for token limit, counter, and strategy. Update example and documentation to reflect new usage patterns. Ensure defaults are applied when options are omitted.

sandyskies commented on model/openai/openai.go:
```diff
@@ -424,6 +458,17 @@ func New(name string, opts ...Option) *Model {
 
+	// Provide defaults at construction time when token tailoring is enabled.
+	// These are best-effort defaults; user-provided counter/strategy always take priority.
+	if o.MaxTokens > 0 {
```
MaxTokens代表的是大模型可以最后输出的token大小
怎么计算我们能输入的token大小呢

Rememorio committed:
refactor: max input tokens

sandyskies commented on examples/tailor/main.go:
```diff
@@ -0,0 +1,268 @@
+	counter := buildCounter(strings.ToLower(*flagCounter), *flagModel, *flagMaxInputTokens)
+	strategy := buildStrategy(counter, strings.ToLower(*flagStrategy))
+	modelInstance := openai.New(*flagModel,
+		openai.WithMaxInputTokens(*flagMaxInputTokens),
```
研究下，其他框架这个行为是默认行为，还是可选的，如果是默认的，我们也需要默认策略

Rememorio replied:
目前的调研结果如下：

- [AutoGen](https://github.com/microsoft/autogen/blob/main/python/packages/autogen-core/src/autogen_core/model_context/_token_limited_chat_completion_context.py#L46): 可选的，支持 middle out 策略。
- [LangChain](https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/messages/utils.py#L715)：可选的，支持 first 和 last 策略。
- [CrewAI](https://github.com/crewAIInc/crewAI/blob/main/src/crewai/utilities/agent_utils.py#L421)：没有暴力裁剪的功能，而是给一个 `respect_context_window` 的 flag，如果为 true 就进行 summary 来缩减输入的 tokens。
- [ADK](https://github.com/google/adk-python), [Agno](https://github.com/agno-agi/agno), [LangGraph](https://github.com/langchain-ai/langgraph) 暂未发现。

这么看来，应该是可选的才对，用户应该不希望自己的 token 被框架默认裁剪掉了，除非他自己手动显式配置。而且这个跟 summary 干的活还有点重叠，直观上来说，使用  summary 来缩减 context window 更加合理，不会丢数据。

sandyskies replied:
这个inputtoken还需要用户手工计算，是不是应该和maxtoken之间有关联？
应该= max context window - max token

sandyskies replied:
可以维护一个映射表，不过内部的模型的话，需要在内部注册和维护

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
refactor(model): use const instead of magic number

Rememorio committed:
feat(model): add model registration and context window resolution functionality

- Implemented functions to register individual and multiple model context window sizes.
- Added tests for model registration, context window resolution, and concurrent access.
- Introduced a new internal model structure to manage context window mappings.

Rememorio committed:
feat(tailor): enhance token tailoring with new flag and automatic management

- Added `-enable-token-tailoring` flag to enable automatic token tailoring based on model context window.
- Updated `max-input-tokens` to allow auto-calculation when set to 0.
- Refactored model initialization to support dual-mode token tailoring.
- Enhanced README with examples for simple and advanced modes of token tailoring.
- Introduced tests for new token tailoring functionality and strategies.

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
feat: update token tailoring example with new features and improvements

- Bump `trpc.group/trpc-go/trpc-agent-go` dependency to v0.3.0.
- Enhance the interactive demo with a new `/show` command to display current messages in head + tail format.
- Improve output messages with emoji indicators for better user experience.
- Refactor token counting logic for consistency across tailoring strategies.
- Update README to reflect new features and usage instructions, including environment variable setup and command descriptions.
- Add tests for calculating preserved message counts in various scenarios.

Rememorio committed:
doc: introduce Token Tailoring feature for efficient message management

- Added Token Tailoring to automatically trim messages exceeding the model's context window limit.
- Implemented dual-mode configuration: automatic and advanced modes for flexible usage.
- Introduced multiple trimming strategies: MiddleOut, HeadOut, and TailOut, with detailed usage examples.
- Included two token counter options: SimpleTokenCounter for speed and TikToken Counter for accuracy.
- Enhanced documentation with core features, quick start guides, and usage examples for better user understanding.

Rememorio committed:
test: add unit tests for calculatePreservedHeadCount and buildPreservedOnlyResult functions

- Introduced tests for calculatePreservedHeadCount to validate the counting of preserved system messages under various scenarios.
- Added tests for buildPreservedOnlyResult to ensure correct message preservation based on specified head and tail counts, including edge cases with tool messages.
- Enhanced test coverage for message handling in token tailoring functionality.

Rememorio commented on model/token_tailor.go:
```diff
@@ -0,0 +1,548 @@
+	result := s.buildMiddleOutResult(messages, preservedHead, bestHeadCount, bestTailCount, preservedTail)
+
+	// Remove the first function execution result message if present.
+	if len(result) > 0 && result[0].Role == RoleTool {
```
已对齐

Rememorio committed:
refactor: remove token tailoring functionality from LLM node

- Deleted the WithTokenTailoring function and related logic from llmRunner.
- Removed associated tests for token tailoring, simplifying the LLM node implementation.
- Adjusted llmRunner structure to eliminate unused fields related to token tailoring.

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

Rememorio committed:
style: format llmRunner comment for improved readability

- Split long comment line in llmRunner struct to enhance clarity and maintain coding standards.

Rememorio committed:
test: add comprehensive unit tests for CountTokens function in SimpleTokenCounter

- Introduced detailed test cases to cover various scenarios in the CountTokens function, including handling of empty messages, basic content, reasoning content, and content parts.
- Ensured that edge cases such as empty reasoning content and non-text content types are properly validated.
- Enhanced test coverage to verify the correct token counting behavior across different message configurations, including unicode characters and combined features.

Rememorio committed:
refactor: simplify token counter initialization and remove max token parameter

- Updated the SimpleTokenCounter to no longer require a maximum token limit during initialization, streamlining its usage across the codebase.
- Adjusted instances in the code where token counters are created, ensuring consistency and reducing complexity.
- Modified related documentation and examples to reflect the changes in token counter instantiation.

Rememorio committed:
test: add unit tests for OpenAI model initialization and token tailoring

- Introduced comprehensive tests for the OpenAI model to verify the initialization of token counters and tailoring strategies based on user-provided configurations.
- Added tests to ensure that default values are correctly assigned when no user configurations are provided, and that concurrent initialization is handled properly.
- Validated that user-defined components take precedence over defaults during initialization.

WineChord commented on docs/mkdocs/zh/model.md:
这里没改，现在应该是不需要入参了。

Rememorio replied:
fixed

WineChord commented on examples/tailor/README.md:
还有这里

Rememorio replied:
fixed

WineChord commented on model/openai/openai.go:
现在有在 limit > 0  的时候自动 enableTokenTailoring = true 吗？

Rememorio replied:
目前没有，已删掉注释

Rememorio committed:
docs: update examples and documentation for SimpleTokenCounter initialization

- Modified examples in English and Chinese documentation to reflect the removal of the maximum token parameter in SimpleTokenCounter initialization.
- Ensured consistency across all instances where SimpleTokenCounter is referenced, simplifying usage instructions for users.

Rememorio committed:
docs: clarify token tailoring requirements in documentation and examples

- Updated English and Chinese documentation to specify that token tailoring is only activated when `WithEnableTokenTailoring(true)` is set.
- Enhanced examples to reflect this requirement, ensuring users understand the necessity of enabling token tailoring for proper functionality.
- Added important notes regarding the distinction between enabling token tailoring and setting maximum input tokens.

Rememorio committed:
Merge branch 'main' into tailor_max_tokens

WineChord reviewed (approved):
LGTM
